test/distributed/org/apache/cassandra/distributed/test/streaming/StreamFailureTest.java

```diff
@@ -0,0 +1,316 @@
+        logger.info("Stream failed logs found: {}", String.join("\n", matches));
+
+        Assertions.assertThat(matches)
+                  .describedAs("node%d expected 1 element but was not true", failingNode.config().num())
+                  .hasSize(1);
+        String logLine = matches.get(0);
+        Assertions.assertThat(logLine).contains(reason);
+
+        Matcher match = Pattern.compile(".*\\[Stream #(.*)\\]").matcher(logLine);
+        if (!match.find()) throw new AssertionError("Unable to parse: " + logLine);
+        UUID planId = UUID.fromString(match.group(1));
+        SimpleQueryResult qr = failingNode.executeInternalWithResult("SELECT * FROM system_views.streaming WHERE id=?", planId);
+        Assertions.assertThat(qr.hasNext()).isTrue();
+        Assertions.assertThat(qr.next().getString("failure_cause")).contains(reason);
+    }
+
+    @Shared
+    public static class TestCondition
+    {
+        private volatile boolean signaled = false;
+
+        public void await()
+        {
+            while (!signaled)
+            {
+                synchronized (this)
+                {
+                    try
+                    {
+                        this.wait();
+                    }
+                    catch (InterruptedException e)
+                    {
+                        throw new AssertionError(e);
+                    }
+                }
+            }
+        }
+
+        public void signal()
+        {
+            signaled = true;
+            synchronized (this)
+            {
+                this.notify();
+            }
+        }
+    }
+
+    @Shared
+    public static class State
+    {
+        public static final TestCondition STREAM_IS_RUNNING = new TestCondition();
+        public static final TestCondition UNBLOCK_STREAM = new TestCondition();
+
+    }
+
+    public static class BB
+    {
+        @SuppressWarnings("unused")
+        public static int writeDirectlyToChannel(ByteBuffer buf, @SuperCall Callable<Integer> zuper) throws Exception
+        {
+            if (isCaller(BigTableZeroCopyWriter.class.getName(), "write"))
+            {
+                State.STREAM_IS_RUNNING.signal();
+                State.UNBLOCK_STREAM.await();
+            }
+            // different context; pass through
+            return zuper.call();
+        }
+
+        @SuppressWarnings("unused")
+        public static boolean append(UnfilteredRowIterator partition, @SuperCall Callable<Boolean> zuper) throws Exception
+        {
+            if (isCaller(CassandraIncomingFile.class.getName(), "read")) // handles compressed and non-compressed
+                throw new java.nio.channels.ClosedChannelException();
+            // different context; pass through
+            return zuper.call();
+        }
+
+        private static boolean isCaller(String klass, String method)
+        {
+            //TODO is there a cleaner way to check this?
```

Comment:
nit: unless you plan to look for a cleaner way in the future, should probably remove the TODO.